utils.py

```diff
@@ -0,0 +1,243 @@
+                #check if validation or testing sites are specified
+                if val_sites and partition == 'trn':
+                    data = data[~data[spatial_idx_name].isin(val_sites)]
+                if test_sites and partition == 'trn':
+                    data = data[~data[spatial_idx_name].isin(test_sites)]
+            # mask out training and test sites from val partition
+            if val_sites and partition == 'val':
+                data = data[data[spatial_idx_name].isin(val_sites)]
+            else:
+                if test_sites and partition=='val':
+                    data = data[~data[spatial_idx_name].isin(test_sites)]
+                if train_sites and partition=='val':
+                    data = data[~data[spatial_idx_name].isin(train_sites)]
+            # mask out training and validation sites from val partition
+            if test_sites and partition == 'tst':
+                data = data[data[spatial_idx_name].isin(test_sites)]
+            else:
+                if train_sites and partition=='tst':
+                    data = data[~data[spatial_idx_name].isin(train_sites)]
+                if val_sites and partition=='tst':
+                    data = data[~data[spatial_idx_name].isin(val_sites)]
+            
+        data.to_csv(filepath)
+
+def RF_model_metrics(pred_obs_csv, spatial_idx_name, time_idx_name,
+                     group_spatially=False, group_temporally=False,
+                     sum_aggregation=False, site_based=False,
+                     outfile=None):
+    data = pd.read_csv(pred_obs_csv)
+    data[time_idx_name] = pd.to_datetime(data[time_idx_name])
+    data.set_index([time_idx_name, spatial_idx_name], inplace=True)
+    
+    var_metrics_list = []
+    
+    if not group_spatially and not group_temporally:
+        metrics = calc_metrics(data)
+        # need to convert to dataframe and transpose so it looks like the
+        # others
+        metrics = pd.DataFrame(metrics).T
+    elif group_spatially and not group_temporally:
+        #note: same as data.groupby(level=spatial_idx_name)
+        metrics = (data.groupby(pd.Grouper(level=spatial_idx_name))
+        .apply(calc_metrics)
+        .reset_index()
+        )
+    elif not group_spatially and group_temporally:
+        if sum_aggregation:
+            #performance metrics computed at the group_temporally timestep
+            #for some reason, no `.` calculations are allowed after .sum(),
+            #so calc_metrics() is called first.
+            if site_based:
+                #create a group_temporally timeseries for each observation site
+                metrics = calc_metrics(data
+                #filter the data to remove nans before computing the sum
+                #so that the same days are being summed in the month.
+                .dropna()
+                .groupby([pd.Grouper(level=time_idx_name, freq=group_temporally),
+                         pd.Grouper(level=spatial_idx_name)])

```

Comment:
This is currently under review in this river-dl PR. This [comment](https://github.com/USGS-R/river-dl/pull/211#discussion_r1149757022) is related. Let me know if you think further explanation of the function arguments is needed.

One edit that may help is stating the number of observations used. When `site_based = True`, the observations are each `group_temporally` timestep for each reach with observations (max of n_segs*n_timesteps). When `site_based = False`, the observations are each `group_temporally` timestep with observations (max of n_timesteps).

Reply:
It's still a gray area to me, but if it is being reviewed else where I will let you two handle it. 

I might suggest a very minimal example/comment documentation could be used to communicate this since two of us are struggling with it.